Source/Initialization/WarpXInitData.cpp

```diff
@@ -307,6 +307,143 @@ WarpX::PostProcessBaseGrids (BoxArray& ba0) const
         AMREX_D_TERM(},},})
         ba0 = BoxArray(std::move(bl));
     }
+
+    bool split_high_density_boxes = false;
+    Real split_high_density_boxes_threshold = 1.1;
+    int split_high_density_boxes_min_box_size = 8;
+    ParmParse pp0;
+    // If there is only one MPI process, we do not split high density boxes,
+    // because the purpose of splitting is to improve load balance potential
+    // among MPI processes.
+    if (ParallelDescriptor::NProcs() > 1) {
+        pp0.queryAdd("warpx.split_high_density_boxes"
+                     ,      split_high_density_boxes);
+        pp0.queryAdd("warpx.split_high_density_boxes_threshold"
+                     ,      split_high_density_boxes_threshold);
+        pp0.queryAdd("warpx.split_high_density_boxes_min_box_size",
+                            split_high_density_boxes_min_box_size);
+    }
+
+    if (split_high_density_boxes) {
+        MultiFab rho;
+        auto const dx = Geom(0).CellSizeArray();
+        auto const problo = Geom(0).ProbLoArray();
+
+        Real wtot = 0; // total number of particles
+
+        std::vector<std::string> species_names;
+        pp0.queryarr("particles.species_names", species_names);
+        for (auto const& species : species_names) { // loop over species
+            Real density_min = std::numeric_limits<amrex::Real>::epsilon();
+            Real nppc = 0;
+            std::string profile, density_function;
+            ParmParse pp_spec(species);
+            pp_spec.query("profile", profile);
+            bool split_using_this_species = false;
+            if (profile == "parse_density_function" &&
+                pp_spec.queryline("density_function(x,y,z)", density_function))
+            {
+                split_using_this_species = true;
+                utils::parser::queryWithParser(pp_spec, "density_min", density_min);
+                std::vector<int> nppc_v(3,1);
+                utils::parser::getArrWithParser(pp_spec, "num_particles_per_cell_each_dim", nppc_v);
```

Comment:
This needs to be generalized to also support `num_particles_per_cell` for `injection_style=NRandomPerCell`

Reply:
Uh, that may be the reason why I haven't seen any effects yet. I'm indeed using `injection_style = NRandomPerCell` with `num_particles_per_cell`